csharp/src/Ice/InputStream.cs

```diff
@@ -2887,17 +2844,55 @@ internal override bool readOptional(int readTag, OptionalFormat expectedFormat)
                 return false;
             }
 
-            private int readInstance(int index, Action<Value> cb)
+            private void Unmarshal(int index, Value v)
+            {
+                //
+                // Add the instance to the map of unmarshaled instances, this must
+                // be done before reading the instances (for circular references).
+                //
+                _unmarshaledMap.Add(index, v);
+
+                //
+                // Read all the deferred indirection tables now that the instance is inserted in _unmarshaledMap.
+                //
+                if (_current.DeferredIndirectionTableList?.Count > 0)
+                {
+                    int savedPos = _stream.pos();
+
+                    Debug.Assert(_current.IndirectionTableList == null || _current.IndirectionTableList.Count == 0);
```

Comment:
Can't this be `Debug.Assert(_current.IndirectionTableList?.Count == 0);`

Reply:
This doesn't work, as null == 0 evaluates to false.